modules/kibana/src/main/java/org/elasticsearch/kibana/KibanaPlugin.java

```diff
@@ -0,0 +1,224 @@
+            return super.prepareRequest(request, new IndexLimitingNodeClient(client, allowedIndexPatterns));
+        }
+    }
+
+    static class IndexLimitingNodeClient extends NodeClient {
+
+        private final NodeClient nodeClient;
+        private final String[] allowedIndexPatterns;
+        private final Automaton allowedIndexAutomaton;
+        private final CharacterRunAutomaton automaton;
+
+        IndexLimitingNodeClient(NodeClient nodeClient, List<String> allowedIndexPatterns) {
+            super(nodeClient.settings(), nodeClient.threadPool());
+            this.nodeClient = nodeClient;
+            this.allowedIndexPatterns = allowedIndexPatterns.toArray(Strings.EMPTY_ARRAY);
+            this.allowedIndexAutomaton = Regex.simpleMatchToAutomaton(this.allowedIndexPatterns);
+            this.automaton = new CharacterRunAutomaton(this.allowedIndexAutomaton);
+        }
+
+        @Override
+        public <Request extends ActionRequest, Response extends ActionResponse> Task executeLocally(ActionType<Response> action,
+                                                                                                    Request request,
+                                                                                                    ActionListener<Response> listener) {
+            final String[] indices;
+            if (request instanceof BulkRequest) {
+                indices = ((BulkRequest) request).requests().stream()
+                    .map(DocWriteRequest::index)
+                    .collect(Collectors.toList())
+                    .toArray(Strings.EMPTY_ARRAY);
+            } else if (request instanceof MultiGetRequest) {
+                indices = ((MultiGetRequest) request).getItems().stream()
+                    .map(Item::index)
+                    .collect(Collectors.toList())
+                    .toArray(Strings.EMPTY_ARRAY);
+            } else if (request instanceof IndicesRequest) {
+                indices = ((IndicesRequest) request).indices();
+            } else {
+                throw new IllegalArgumentException("This client cannot be used to make a non indices request");
+            }
+
+            if (indices == null || indices.length == 0 ||
+                (indices.length == 1 && (IndexNameExpressionResolver.isAllIndices(List.of(indices)) || indices[0].equals("*")))) {
+                if (request instanceof IndicesRequest.Replaceable) {
+                    // just replace this with the allowed system index patterns
+                    ((IndicesRequest.Replaceable) request).indices(allowedIndexPatterns);
+                } else {
+                    throw new IllegalStateException("Unable to replace indices on request " +
+                        request.getClass().getSimpleName());
+                }
+            } else {
+                for (String index : indices) {
+                    // TODO this will not handle date math, is that OK? Do we need wildcard support?
```

Comment:
We might need wildcard support eventually, although I don't *think* we do for Kibana (@tylersmalley do you know?). Date math support is more questionable. It might be worth trying to detect it and throwing an exception, similar to what we do here.